### PR TITLE
Fit plane constructor

### PR DESCRIPTION
**Pull request by zerlinazl**
Created constructor and replaced xz _fit_from_photobleached_lines() function with a stub for xy

### CODE REVIEW

File: plane/fit_plane.py [hunk 1]
```diff
@@ -5,71 +5,54 @@
...
+        This function initializes a FitPlane by a list of points for template centers.
+
         INPUTS:
-            fluorescence_image_points_on_line_pix: For each of the photobleach lines, 
-                find at least two points in the fluorescence image. Mark the coordinates as pixel values
-                l1 = [[x1,y1],[x2,y2],...] and create an array of those [l1,l2,...,ln]
-            photobleach_line_position_mm: an array defining the position (in mm) of each of the photobleach line positions 
-            photobleach_line_group: an array defining each line is a horizontal or vertical line ['h','v',...] etc
+            template_center_positions_uv_pix: For each photobleach barcode, find the center position in pixels. This is an
+                array of these center points.
```
Comment (MyYo) [inline plane/fit_plane.py:23]: Can you give an example? Is it numpy.zeros(n,2) or numpy.zeros(2,n) where the "2" dimension is u,v and n is number of points?

File: plane/fit_plane.py [hunk 2]
```diff
@@ -5,71 +5,54 @@
...
+            template_center_positions_xyz_um: An array of points defining the position (in um) of the locations that each of
```
Comment (MyYo) [inline plane/fit_plane.py:24]: Can you give an example? Is it numpy.zeros(n,3) or numpy.zeros(3,n) where the "3" dimension is x,y,z and n is number of points?

File: plane/fit_plane.py [hunk 3]
```diff
@@ -5,71 +5,54 @@
...
         # Input check
-        if (len(fluorescence_image_points_on_line_pix) != len(photobleach_line_position_mm) or
-            len(fluorescence_image_points_on_line_pix) != len(photobleach_line_group)):
-            raise ValueError("Number of lines should be the same between " + 
-                "fluorescence_image_points_on_line_pix, photobleach_line_position_mm, photobleach_line_group")
+        if (len(template_center_positions_uv_pix) != len(template_center_positions_xyz_um)):
+            raise ValueError("Number of points should be the same between " + 
+                "template_center_positions_uv_pix, template_center_positions_xyz_um")
         
         # Solve x,y first
```
Comment (MyYo) [inline plane/fit_plane.py:42]: Comment "solve x,y first" here doesn't make sense in this context

File: plane/fit_plane.py [hunk 4]
```diff
@@ -86,50 +69,14 @@ def from_json(cls, json_str):
...
-        self.h = np.array([x[4], x[5], np.nan])
+    """ End constructor methods """    
+    def _fit_from_templates(self, template_center_positions_uv_pix, template_center_positions_xyz_um):
+        """
+        Calculate a mapping with vectors u, v, h to project points from uv coordinates to xyz physical locations.
+        """
+        pass
         
-    def _fit_from_photobleach_lines_z_from_no_shear_equal_size(self):
+    def _fit_from_templates_z_from_no_shear_equal_size(self):
```
Comment (MyYo) [inline plane/fit_plane.py:79]: After reading it here again. You can remove this check - it doesn't make sense

**Review by MyYo — CHANGES_REQUESTED** [review]
Thanks, left a few comments.

A general ask is to add the constructor to to a tester

File: plane/fit_plane.py (hunk 4 above)
Comment (zerlinazl) [reply plane/fit_plane.py:79]: Remove the whole function?

Comment (MyYo) [reply plane/fit_plane.py]: Yes the whole function

File: plane/fit_plane.py [hunk 5]
```diff
@@ -5,71 +5,55 @@
...
+        This function initializes a FitPlane by a list of points for template centers.
+
         INPUTS:
-            fluorescence_image_points_on_line_pix: For each of the photobleach lines, 
-                find at least two points in the fluorescence image. Mark the coordinates as pixel values
-                l1 = [[x1,y1],[x2,y2],...] and create an array of those [l1,l2,...,ln]
-            photobleach_line_position_mm: an array defining the position (in mm) of each of the photobleach line positions 
-            photobleach_line_group: an array defining each line is a horizontal or vertical line ['h','v',...] etc
+            template_center_positions_uv_pix: For each photobleach barcode, find the center position in pixels. This is an
+                array of these center points [(x1, y1), (x2, y2),..., (xn, yn)] with shape (n,2)
```
Comment (MyYo) [inline plane/fit_plane.py:23]: I'm confused, is the input a list of tuples?  I recommend changing to a numpy matrix of shape (n,2)

File: plane/fit_plane.py [hunk 6]
```diff
@@ -5,71 +5,55 @@
...
+            raise ValueError("Number of points should be the same between " + 
+                "template_center_positions_uv_pix, template_center_positions_xyz_um")
         
         # Solve x,y first
-        c = fp._fit_from_photobleach_lines_xy(
-            fluorescence_image_points_on_line_pix, 
-            photobleach_line_position_mm, 
-            photobleach_line_group)
-            
+        fp._fit_from_templates(
```
Comment (MyYo) [inline plane/fit_plane.py:44]: The comment in row 43 makes no sense...

File: plane/fit_plane.py [hunk 7]
```diff
@@ -5,71 +5,55 @@
...
-        # c_u - according to the location that norm hits the plane
-        # c_v - center of the fluorescence_image_points_on_line_pix
-        v_coordinates = np.array([item[1] for sublist in fluorescence_image_points_on_line_pix for item in sublist])
-        c_v = np.mean(v_coordinates)
-        o = fp.get_uv_from_xyz([0,0,0])
-        c_u = o[0]
-        fp.recommended_center_pix = np.array([c_u, c_v])
-        
-        return(fp)
+        if fp.u:
```
Comment (MyYo) [inline plane/fit_plane.py:50]: Remove "Fix z component" it's not relevant in this case

File: plane/test_fit_plane.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+from plane.fit_plane import FitPlane
+
+class TestFitMultiPlane(unittest.TestCase):
+
+    def setUp(self):
+      self.template_center_positions_uv_pix = [[0,1],[1,0],[1,1]]
+      self.template_center_positions_xyz_um = [[0,2,10], [2,0,10], [2,2,11]]
+
+    def test_main_function_runs(self):
+        FitPlane.from_template_centers(self.template_center_positions_uv_pix, self.template_center_positions_xyz_um, print_inputs=False)
```
Comment (MyYo) [inline plane/test_fit_plane.py:15]: Test with print_input=True to check that it works

**Review by MyYo — CHANGES_REQUESTED** [review]

File: plane/fit_plane.py (hunk 5 above)
Comment (zerlinazl) [reply plane/fit_plane.py]: sorry, was supposed to be brackets. It now accepts either list or numpy array and converts to numpy

**Review by MyYo — APPROVED** [review]
Thanks looks good to me!

**Pull request merged**